Runtime init the cleaner instance in ResourceCleaner

jerboaa opened the pull request:
Inject an accessor for the ResourceCleaner.CLEANER static variable so that the associated cleaner won't get initialized at build time and ends up being in the native image heap.

Closes: #31440

jerboaa commented:
Mandrel workflow with this is running here: https://github.com/graalvm/mandrel/actions/runs/4285051293

jerboaa commented:
It should probably be in the resteasy-common extension rather than resteasy-client as I had it before.

jerboaa commented:
@zakkak Please take a look when you get a chance! Thanks.

zakkak commented on extensions/resteasy-classic/resteasy-common/runtime/src/main/java/io/quarkus/resteasy/common/runtime/graal/ResourceCleanerReplacement.java:
```diff
@@ -0,0 +1,45 @@
+        }
+
+        private static synchronized Cleaner initializeOnce() {
+            Cleaner result = INSTANCE;
+            if (result == null) {
+                // double checked locking is ok as INSTANCE is volatile
+                result = INSTANCE = Cleaner.create();
+            }
+            return result;
```
Styling nitpick

```suggestion
            if (INSTANCE == null) {
                // double checked locking is ok as INSTANCE is volatile
                INSTANCE = Cleaner.create();
            }
            return INSTANCE;
```

jerboaa replied:
Done.

zakkak reviewed (commented):
LGTM, but did you try just reinitializing the class instead?

jerboaa commented:
> ... but did you try just reinitializing the class instead?

I did not. Then again, wouldn't for runtime-reinited classes the same restrictions apply (instances of those classes are disallowed in the image heap)? I'll have a look and will report back.

zakkak commented:
> > ... but did you try just reinitializing the class instead?
> 
> I did not. Then again, wouldn't for runtime-reinited classes the same restrictions apply (instances of those classes are disallowed in the image heap)? 

No, fields of reinited classes are treated differently. The issue with instances of certain classes being present in the native heap is that those instances were generated at build time and they are thus not valid at runtime. When reiniting a class the said fields are "uninitialized" in the initial native heap and get initialized at run time which is valid.

jerboaa committed:
Runtime init resteasy's ResourceCleaner

ResourceCleaner contains a static reference to a java.lang.ref.Cleaner
instance which are disallowed in the native image heap. Runtime
initializing that class instead fixes this issue.

Closes: #31440

zakkak commented:
> I thought this is for **runtime** initialized classes. Runtime **re-inited** classes are initialized both at build time and again at runtime: https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/classinitialization/InitKind.java#L44..L45
> 

Correct, they are initialized at build time for the build to proceed and to increase the level of optimization, but the values from the static fields are not used at runtime (if I am not mistaken they are not even placed in the generated native heap), instead they get re-initialized at run time. Because of this they are allowed to be _present_ in the heap.

> Anyway, in this case both seem to work.

I didn't suggest pure runtime initialization because I thought the class needed to be build-time initialized. Great that you tried both.

> Runtime-reinit: https://github.com/quarkusio/quarkus/compare/main...jerboaa:quarkus:runtime-reinit-resource?expand=1 Runtime-init: https://github.com/quarkusio/quarkus/compare/main...jerboaa:quarkus:runtime-init-resource?expand=1
> 
> Shall I update this patch with the last (runtime-init) option?

Yes, please.

zakkak commented:
> Runtime-init: https://github.com/quarkusio/quarkus/compare/main...jerboaa:quarkus:runtime-init-resource?expand=1

@jerboaa I suggest doing the runtime initialization registration in https://github.com/quarkusio/quarkus/blob/main/extensions/resteasy-classic/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java instead of creating a new class.

jerboaa commented:
@zakkak Done now. Thanks for the reviews. I've also rescheduled a run of the mandrel tests with this:
https://github.com/graalvm/mandrel/actions/runs/4285051293

zakkak reviewed (approved):
LGTM, thanks!

jerboaa commented:
@gsmet Hi! Could you please help approving the workflow and getting this integrated? It fixes many broken integration tests on the mandrel CI side. Thanks!

gsmet commented:
@zakkak just to be sure, you have the permission to approve a workflow run, right? Wondering if it's only admins.

zakkak commented:
> @zakkak just to be sure, you have the permission to approve a workflow run, right? Wondering if it's only admins.

Yes, I was the one who approved it this morning. I am also going to merge this once the CI completes. @jerboaa was probably not aware that I have the power to do both :)

Probably it didn't help that I didn't notice that the workflow was blocked for some unknown (since this is not the first contribution of @jerboaa) reason.

zakkak commented:
The failure is unrelated to the PR. Merging it.

zakkak commented:
Marking this for backport since although [the Cleaner instances were disallowed only recently](https://github.com/oracle/graal/pull/6062) the issue is actually present in older versions as well.

gsmet commented:
Actually, AFAICS, `ResourceCleaner` doesn't exist in 2.16.

zakkak commented:
Indeed, the ResourceCleaner was [added in 6.1.0.Final](https://github.com/resteasy/resteasy/commit/4dbae7119a3966a244cfe1629077cef29ea2fd8f) and 2.16 is [using 4.7.7](https://github.com/quarkusio/quarkus/blob/0865c6a050c07e8543582cc9bb7e5c6b82593602/bom/application/pom.xml#L22).
